src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+pub struct ValueShare<W: VidpfWeight> {
+    /// Secret share of the input's weight.
+    pub share: W,
+    /// Proof used to verify the share.
+    pub proof: Proof,
+}
+
+/// Proof size in bytes.
+const VIDPF_PROOF_SIZE: usize = 32;
+
+/// [`Proof`] is used to validate user input and shares after evaluation.
+#[derive(Debug, PartialEq, Eq, Default)]
+pub struct Proof([u8; VIDPF_PROOF_SIZE]);
+
+impl Proof {
+    fn xor(mut self, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.bitxor_assign(b));
+        self
+    }
+
+    fn conditional_xor(mut self, choice: Choice, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.conditional_assign(&a.bitxor(b), choice));
+        self
+    }
+}
+
+/// [`Seed`] is used to feed a pseudorandom generator during evaluation.
+type Seed = [u8; 16];
+
+struct PrgOutput {
+    left_seed: Seed,
+    left_control_bit: Choice,
+    right_seed: Seed,
+    right_control_bit: Choice,
+}
+
+/// [`Weight`] represents an array of field elements that implements the [`VidpfWeight`] trait.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Weight<F: FieldElement>(Vec<F>);
+
+impl<F: FieldElement> Weight<F> {
+    /// Creates a [`Weight`] from an array of field elements with length specified in parameters.
+    pub fn new_with_params(params: &WeightParameter<F>, array: &[F]) -> Result<Self, VidpfError> {
+        if params.length != array.len() {
+            return Err(VidpfError::InvalidWeightLength);
+        }
+
+        let slice = array
+            .get(0..params.length)
+            .ok_or(VidpfError::InvalidWeightLength)?;
+
+        Ok(Self(Vec::from(slice)))
+    }
+
+    /// Panics if weight lengths are different.
+    fn check_eq_length(a: &Self, b: &Self) {
+        if a.0.len() != b.0.len() {
+            panic!("vidpf: weights with different number of field elements")
+        }
+    }
+}
+
+/// [`WeightParameter`] contains the run-time parameters to create an array of field elements.
+#[derive(Debug, Clone, Copy)]
+pub struct WeightParameter<F: FieldElement> {
+    /// The number of field elements.
+    pub length: usize,
+    /// Parameters used to create field elements.
+    pub field_parameter: <F as IdpfValue>::ValueParameter,
+}
+
+impl<F: FieldElement> VidpfWeight for Weight<F> {}
+
+impl<F: FieldElement> IdpfValue for Weight<F> {
+    type ValueParameter = WeightParameter<F>;
+
+    fn generate<S: RngCore>(seed_stream: &mut S, parameter: &Self::ValueParameter) -> Self {
+        Self(
+            (0..parameter.length)
+                .map(|_| <F as IdpfValue>::generate(seed_stream, &parameter.field_parameter))
+                .collect(),
+        )
+    }
+
+    fn zero(parameter: &Self::ValueParameter) -> Self {
+        Self(
+            (0..parameter.length)
+                .map(|_| <F as IdpfValue>::zero(&parameter.field_parameter))
+                .collect(),
+        )
+    }
```

Comment:
An explicit parameter would simplify this code substantially.
```suggestion
impl<F: FieldElement> VidpfWeight for Weight<F> {}

impl<F: FieldElement> IdpfValue for Weight<F> {
    type ValueParameter = usize;

    fn generate<S: RngCore>(seed_stream: &mut S, length: &usize) -> Self {
        Self(
            (0..*length)
                .map(|_| <F as IdpfValue>::generate(seed_stream, &()))
                .collect(),
        )
    }

    fn zero(length: &usize) -> Self {
        Self((0..*length).map(|_| F::zero()).collect())
    }

    fn conditional_select(lhs: &Self, rhs: &Self, choice: Choice) -> Self {
        Self::check_eq_length(lhs, rhs);
        Self(
            zip(&lhs.0, &rhs.0)
                .map(|(a, b)| <F as IdpfValue>::conditional_select(a, b, choice))
                .collect(),
        )
    }
}
```

Reply:
it does, thanks.